nikola/nikola.py

```diff
@@ -1062,30 +1081,47 @@ def init_plugins(self, commands_only=False, load_all=False):
             self.plugin_manager._candidates = list(set(self.plugin_manager._candidates) - bad_candidates)
 
-        # Find repeated plugins and discard the less local copy
-        def plugin_position_in_places(plugin):
-            # plugin here is a tuple:
-            # (path to the .plugin file, path to plugin module w/o .py, plugin metadata)
-            for i, place in enumerate(self._plugin_places):
-                if plugin[0].startswith(place):
-                    return i
+        self.plugin_manager._candidates = self._filter_duplicate_plugins(self.plugin_manager._candidates)
+        self.plugin_manager.loadPlugins()
 
-        plugin_dict = defaultdict(list)
-        for data in self.plugin_manager._candidates:
-            plugin_dict[data[2].name].append(data)
-        self.plugin_manager._candidates = []
-        for name, plugins in plugin_dict.items():
-            if len(plugins) > 1:
-                # Sort by locality
-                plugins.sort(key=plugin_position_in_places)
-                utils.LOGGER.debug("Plugin {} exists in multiple places, using {}".format(
-                    plugins[-1][2].name, plugins[-1][0]))
-            self.plugin_manager._candidates.append(plugins[-1])
+        # Search for compiler plugins which we disabled but shouldn't have
+        self._activate_plugins_of_category("PostScanner")
+        if not load_all:
+            file_extensions = set()
+            for post_scanner in [p.plugin_object for p in self.plugin_manager.getPluginsOfCategory('PostScanner')]:
+                exts = post_scanner.supported_extensions()
+                if exts is not None:
+                    file_extensions.update(exts)
+                else:
+                    # Stop scanning for more: once we get None, we have to load all compilers anyway
+                    file_extensions = None
+                    break
+            to_add = []
+            for k, v in compilers.items():
+                if file_extensions is None or file_extensions.intersection(v):
+                    self.config['COMPILERS'][k] = sorted(list(v))
+                    self.bad_compilers.remove(k)
+                    p = self.disabled_compilers.pop(k, None)
+                    if p:
+                        to_add.append(p)
+                    for p in self.disabled_compiler_extensions.pop(k, []):
+                        to_add.append(p)
+            for _, p in self.disabled_compilers.items():
+                utils.LOGGER.debug('Not loading unneeded compiler {}', p[-1].name)
+            for _, plugins in self.disabled_compiler_extensions.items():
+                for p in plugins:
+                    utils.LOGGER.debug('Not loading compiler extension {}', p[-1].name)
+            if to_add:
+                self.plugin_manager._candidates = self._filter_duplicate_plugins(to_add)
+                self.plugin_manager.loadPlugins()
```

Comment:
Yes. Until that point, only plugins of type `PostScanner` have been activated, and all others not. Since we only added plugins of type `PageCompiler` or compiler extensions (plugins having `compiler` set) no plugin should be forgotten.

(Except if a post scanner plugin has `compiler` set in its `.plugin` file. I assumed such plugins don't exist.)